Support disabling the sbt task linter

eatkins opened the pull request:
- [x] I've read the [CONTRIBUTING](https://github.com/sbt/sbt/blob/develop/CONTRIBUTING.md) guidelines
This PR resolves #3266. It also does some minor clean up of the TaskLinterDSL to make it more maintainable in general.

eatkins committed:
Switch to more functional style in BaseTaskLinterDSL

I found it somewhat difficult to reason about the state of the tree
Traverser because of the usage of mutable variables and data structures.
For example, I determined that the uncheckedWrappers were never used
non-locally so a Set didn't seem to be the right data structure. It was
reasonably straightforward to switch to a more functional style by
parameterizing the method local traverser class.

Bonus:
  - remove unused variable disableNoValueReport
  - add scaladoc to document the input parameters of the traverser class
    so that it's a bit easier to understand for future maintainers.

eatkins committed:
Lint TaskLinterDSL for intellij

This cleans up all of the yellow intellij warnings for me. It still
complains about some typos, but those manifest as green squiggled lines
which are less annoying.

eatkins committed:
Add solutions for dynamic task evaluation

There are many cases where one would want to force evaluation of the
task even when contained in a lambda (see
https://github.com/sbt/sbt/issues/3266). The @sbtUnchecked annotation is
one way to disable the linter that prevents this, but it is obscure. If
the annotation is to exist, I think it should be presented as a
solution.

eatkins committed:
Make the task linter configurable

The user should be able to configure whether or not the task linting is
strictly enforced. In my opinion, the linter is generally pretty good
about warning users that a particular definition may not behave the way
the user expects. Unfortunately, it is fairly common that the user
explicitly wants this behavior and making the linter abort compilation
is a frustrating user experience.

To fix this, I add the LinterLevel trait to a new sbt.dsl package in the
core-macros project. The user can configure the linter to:
1) abort when an issue is detected. This is the current default behavior.
2) print a warning when an issues is detected
3) skip linting all together

Linter configuration is managed by importing the corresponding implicit
object from the LinterLevel companion object.

Fixes #3266

eatkins commented:
cc @dwijnand @eed3si9n. I whipped this up fairly quickly and didn't give a ton of thought to the names and project layout.

dwijnand commented:
I reviewed it _really_ really quick and it looked good, but I also approve of fixing that issue in general.  Thanks, Ethan.

eatkins commented on main-settings/src/main/scala/sbt/dsl/LinterLevel.scala:
```diff
@@ -0,0 +1,17 @@
+
+package sbt.dsl
+
+sealed trait LinterLevel
```
I considered contraband for this, but wasn't sure if there is a way to put the (prioritized) implicit instances in the companion object.

eed3si9n replied:
You can use https://www.scala-sbt.org/contraband/schema.html#Extra+code

```
  #xcompanioninterface CompanionInterface1
```

eed3si9n commented:
This is awesome. Thanks @eatkins!

eed3si9n commented on main-settings/src/main/scala/sbt/dsl/LinterLevel.scala:
```diff
@@ -0,0 +1,17 @@
+
+sealed trait LinterLevel
+object LinterLevel extends LinterLevelLowPriority {
+  implicit case object Warn extends LinterLevel
```
There should be an explanation of why this is made implicit, and how this is meant to be used.

eatkins replied:
I force pushed an update that added scaladoc to these.

eatkins committed:
Fix Def cannot be used inside a taskDyn #3110

The illegalReference check did not actually validate whether the illegal
reference actually referred to an M[_] (which is pretty much always
Initialize[_]]). The means by which this failure was induces were fairly
obscure and go through multiple levels of macro transformations that I
attempt to explain in the comment in IllegalReferenceSpec.

Fixes #3110

eatkins committed:
Make the TaskLinterDSL warn by default

I am generally of the opinion that a linter should not abort progress by
default. I do, however, think that it should be on by default, making
warn a happy compromise.

dwijnand commented:
Where did that fix come from?!  Very nice.

If you want a more common source of dynamic tasks put some projects in a seq and then map the seq, scope a key and .value that.  Users try to do that all the time.

```scala
Seq(foo, bar, baz).map(p =>
  (bark in p).value
)
```

eatkins commented:
Funnily enough, this is all related #3718. I'm working on trying to implement the taskSources idea in the comment at the bottom. This requires traversing the list of dependencies for each task that is being watched to extract all of the taskSources dependencies. While working on that, I realized that dynamic tasks are a huge problem because we lose the dependency information for dynamic tasks. I started working on modifying the Def.taskDyn macro to try and capture all of the dependencies for the Def.initialize instances that are referenced in the dynamic task definition. This forced me to get quite familiar with the TaskLinterDSL at which point I decided to go ahead and fix #3110 and #3266, both of which had been irritating me for a long time.
